### ISSUE TITLE
Replace TimeInterval in APIs with DispatchTimeInterval

### ISSUE THREAD

**Issue by NachoSoto** [ReactiveCocoa/ReactiveSwift#33]
## Reasoning:

The current `TimeInterval` APIs are fragile and dangerous for many reasons:
- We always end up converting to milliseconds, but if the input value is too high, it will overflow (#31 fixed this to some extent by reducing the likelihood, but it's still possible).
- The API is not strongly typed: there is no guarantee or way to enforce users sending the right magnitude (seconds versus milliseconds, etc), other than a weak comment. 

But with `Swift 3` we can do better: using `DispatchTimeInterval` solves all these problems :)
## Process:

So I'm proposing deprecating current APIs that use `TimeInterval` to point users to use new `DispatchTimeInterval` versions.

**Comment by ikesyo** [comment]
We probably need an extension like this:

```swift
extension Date {
    internal func addingDispatchTimeInterval(_ interval: DispatchTimeInterval) -> Date {
        switch interval {
        case let .seconds(s): ...
        case let .milliseconds(ms): ...
        ....
        }
    }
}
```

**Comment by liscio** [comment]
> We probably need an extension like this:

Should we not instead remove the use of `Date` entirely in favour of `DispatchTime`? I never did like using `{NS,}Date` to represent a point in time…

The usage gets much nicer, too. Compare using `Date().addingDispatchTimeInterval(.milliseconds(300))` vs. `.now() + .milliseconds(300)`, for example.

**Comment by NachoSoto** [comment]
Yup 👍

**Comment by liscio** [comment]
I've gone ahead and done much of the work already here for a whole-hog replacement of `Date` and `TimeInterval`: https://github.com/ReactiveCocoa/ReactiveSwift/compare/master...liscio:cl-dispatchwalltime

Unfortunately, it appears that `DispatchWallTime` (and `DispatchTime`, but we don't want to base our scheduler on that) is busted in the Swift standard library. I filed a bug [SR-2807](https://bugs.swift.org/browse/SR-2807) to track the issues I encountered. Specifically, the crash when comparing two `DispatchWallTime`s is going to block progress.

**Comment by liscio** [comment]
I've *also* gone ahead and opened #48 with the approach that was originally proposed, since it's not practical to expect the above issue to get corrected in time for 1.0.

**Comment by ashfurrow** [comment]
It looks like this was fixed by #48 (good job!). Is the 1.0 milestone up to date? This is the only open issue left in it at the moment.

**Comment by mdiep** [comment]
I just added #88 to the 1.0 milestone because #22 was in there. I think those are the last things that we need to resolve.
